In [1]:
# Importing Necessary Libraries
import pandas as pd 
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
# Importing Dataset
data = pd.read_csv("salary.csv")
pd.set_option("display.max_columns", None)
data

C:\Users\sweet\AppData\Local\Temp\ipykernel_13796\3484571091.py:2: DtypeWarning: Columns (16,18,20) have mixed types. Specify dtype option on import or set low_memory=False.
  data = pd.read_csv("salary.csv")


AGENCY                                        AGENCY NAME  \
0          241  COMPTROLLER OF PUBLIC ACCOUNTS, JUDICIARY SECT...   
1          212  OFFICE OF COURT ADMINISTRATION                ...   
2          241  COMPTROLLER OF PUBLIC ACCOUNTS, JUDICIARY SECT...   
3          212  OFFICE OF COURT ADMINISTRATION                ...   
4          696  TEXAS DEPARTMENT OF CRIMINAL JUSTICE          ...   
...        ...                                                ...   
149476     809  STATE PRESERVATION BOARD                      ...   
149477     809  STATE PRESERVATION BOARD                      ...   
149478     809  STATE PRESERVATION BOARD                      ...   
149479     809  STATE PRESERVATION BOARD                      ...   
149480     809  STATE PRESERVATION BOARD                      ...   

                             LAST NAME                      FIRST NAME MI  \
0       RUCKER                          MORTON                          V   
1       RUCKER                          MORTON                          V   
2       SPECIA JR                       JOHN                            J   
3       SPECIA JR                       JOHN                            J   
4       ONTIVEROS                       ESTHER                              
...                                ...                             ... ..   
149476  WESSELS                         JOHN                            P   
149477  WINDHAM                         EVAN                            A   
149478  WRIGHT                          DERRICK                         C   
149479  YOUNG                           DOUGLAS                         R   
149480  ZUNKER                          GEORGIA                         P   

       CLASS CODE                                        CLASS TITLE  \
0        JD25      JUDGE, RETIRED                                ...   
1        3524      GENERAL COUNSEL IV                            ...   
2        JD25      JUDGE, RETIRED                                ...   
3        3524      GENERAL COUNSEL IV                            ...   
4        4504      CORREC  OFFICER IV                            ...   
...           ...                                                ...   
149476   6232      SECURITY OFFICER III                          ...   
149477   0302      WEB ADMINISTRATOR III                         ...   
149478   0130      CUSTOMER SERVICE REP I                        ...   
149479   1572      PROGRAM SPECIALIST III                        ...   
149480   0130      CUSTOMER SERVICE REP I                        ...   

              ETHNICITY           GENDER  \
0       WHITE            MALE              
1       WHITE            MALE              
2       WHITE            MALE              
3       WHITE            MALE              
4       HISPANIC         FEMALE            
...                 ...              ...   
149476  WHITE            MALE              
149477  WHITE            FEMALE            
149478  WHITE            MALE              
149479  WHITE            MALE              
149480  WHITE            FEMALE            

                                          STATUS EMPLOY DATE  HRLY RATE  \
0       URP - UNCLASSIFIED REGULAR PART-TIME        02/18/88   75.96150   
1       CTP - CLASSIFIED TEMPORARY PART-TIME        02/01/15   81.04454   
2       URP - UNCLASSIFIED REGULAR PART-TIME        02/01/20   75.96150   
3       CTP - CLASSIFIED TEMPORARY PART-TIME        09/01/18   81.04453   
4       CRF - CLASSIFIED REGULAR FULL-TIME          06/29/20    0.00000   
...                                          ...         ...        ...   
149476  CRF - CLASSIFIED REGULAR FULL-TIME          10/30/17    0.00000   
149477  CRF - CLASSIFIED REGULAR FULL-TIME          07/13/15    0.00000   
149478  CRP - CLASSIFIED REGULAR PART-TIME          10/15/12   12.93000   
149479  CRF - CLASSIFIED REGULAR FULL-TIME          09/22/89    0.00000   
149480  CRP - CLASSIFIED REGULAR PA

In [3]:
# There is a warning while loading the dataset this is due to the dataset contains mixed data types to overcome this
# Loading the dataset with Low_memory set to False
data = pd.read_csv("salary.csv", low_memory= False)
data

AGENCY                                        AGENCY NAME  \
0          241  COMPTROLLER OF PUBLIC ACCOUNTS, JUDICIARY SECT...   
1          212  OFFICE OF COURT ADMINISTRATION                ...   
2          241  COMPTROLLER OF PUBLIC ACCOUNTS, JUDICIARY SECT...   
3          212  OFFICE OF COURT ADMINISTRATION                ...   
4          696  TEXAS DEPARTMENT OF CRIMINAL JUSTICE          ...   
...        ...                                                ...   
149476     809  STATE PRESERVATION BOARD                      ...   
149477     809  STATE PRESERVATION BOARD                      ...   
149478     809  STATE PRESERVATION BOARD                      ...   
149479     809  STATE PRESERVATION BOARD                      ...   
149480     809  STATE PRESERVATION BOARD                      ...   

                             LAST NAME                      FIRST NAME MI  \
0       RUCKER                          MORTON                          V   
1       RUCKER                          MORTON                          V   
2       SPECIA JR                       JOHN                            J   
3       SPECIA JR                       JOHN                            J   
4       ONTIVEROS                       ESTHER                              
...                                ...                             ... ..   
149476  WESSELS                         JOHN                            P   
149477  WINDHAM                         EVAN                            A   
149478  WRIGHT                          DERRICK                         C   
149479  YOUNG                           DOUGLAS                         R   
149480  ZUNKER                          GEORGIA                         P   

       CLASS CODE                                        CLASS TITLE  \
0        JD25      JUDGE, RETIRED                                ...   
1        3524      GENERAL COUNSEL IV                            ...   
2        JD25      JUDGE, RETIRED                                ...   
3        3524      GENERAL COUNSEL IV                            ...   
4        4504      CORREC  OFFICER IV                            ...   
...           ...                                                ...   
149476   6232      SECURITY OFFICER III                          ...   
149477   0302      WEB ADMINISTRATOR III                         ...   
149478   0130      CUSTOMER SERVICE REP I                        ...   
149479   1572      PROGRAM SPECIALIST III                        ...   
149480   0130      CUSTOMER SERVICE REP I                        ...   

              ETHNICITY           GENDER  \
0       WHITE            MALE              
1       WHITE            MALE              
2       WHITE            MALE              
3       WHITE            MALE              
4       HISPANIC         FEMALE            
...                 ...              ...   
149476  WHITE            MALE              
149477  WHITE            FEMALE            
149478  WHITE            MALE              
149479  WHITE            MALE              
149480  WHITE            FEMALE            

                                          STATUS EMPLOY DATE  HRLY RATE  \
0       URP - UNCLASSIFIED REGULAR PART-TIME        02/18/88   75.96150   
1       CTP - CLASSIFIED TEMPORARY PART-TIME        02/01/15   81.04454   
2       URP - UNCLASSIFIED REGULAR PART-TIME        02/01/20   75.96150   
3       CTP - CLASSIFIED TEMPORARY PART-TIME        09/01/18   81.04453   
4       CRF - CLASSIFIED REGULAR FULL-TIME          06/29/20    0.00000   
...                                          ...         ...        ...   
149476  CRF - CLASSIFIED REGULAR FULL-TIME          10/30/17    0.00000   
149477  CRF - CLASSIFIED REGULAR FULL-TIME          07/13/15    0.00000   
149478  CRP - CLASSIFIED REGULAR PART-TIME          10/15/12   12.93000   
149479  CRF - CLASSIFIED REGULAR FULL-TIME          09/22/89    0.00000   
149480  CRP - CLASSIFIED REGULAR PA

# 1) Basic Checks and some data filtering

In [5]:
# Let's rename the columns first for our convenience
# Define a dictionary with current column names (all caps) as keys and new names as values
new_column_names = {
    'AGENCY': 'agency',
    'AGENCY NAME': 'agency_name',
    'LAST NAME': 'last_name',
    'FIRST NAME': 'first_name',
    'MI': 'middle_initial',
    'CLASS TITLE': 'class_title',
    'CLASS CODE': 'class_code',
    'ETHNICITY': 'ethnicity',
    'GENDER': 'gender',
    'STATUS': 'status',
    'EMPLOY DATE': 'employ_date',
    'HRLY RATE': 'hourly_rate',
    'HRS PER WK': 'hours_per_week',
    'MONTHLY': 'monthly_income',
    'ANNUAL': 'annual_income',
    'summed_anual_salary': 'summed_annual_income',
    'STATE NUMBER': 'state_number',
    'duplicated': 'is_duplicated',
    'combined_multiple_jobs': 'combined_multiple_jobs',
    'multiple_full_time_jobs': 'multiple_full_time_jobs',
    'hide_from_search': 'hide_from_search'
}

# Rename the columns
data.rename(columns=new_column_names, inplace=True)

# Display the new column names to verify
print(data.columns)

Index(['agency', 'agency_name', 'last_name', 'first_name', 'middle_initial',
       'class_code', 'class_title', 'ethnicity', 'gender', 'status',
       'employ_date', 'hourly_rate', 'hours_per_week', 'monthly_income',
       'annual_income', 'state_number', 'is_duplicated',
       'multiple_full_time_jobs', 'combined_multiple_jobs',
       'summed_annual_salary', 'hide_from_search'],
      dtype='object')


In [6]:
data.head()

agency                                        agency_name  \
0     241  COMPTROLLER OF PUBLIC ACCOUNTS, JUDICIARY SECT...   
1     212  OFFICE OF COURT ADMINISTRATION                ...   
2     241  COMPTROLLER OF PUBLIC ACCOUNTS, JUDICIARY SECT...   
3     212  OFFICE OF COURT ADMINISTRATION                ...   
4     696  TEXAS DEPARTMENT OF CRIMINAL JUSTICE          ...   

                        last_name                      first_name  \
0  RUCKER                          MORTON                           
1  RUCKER                          MORTON                           
2  SPECIA JR                       JOHN                             
3  SPECIA JR                       JOHN                             
4  ONTIVEROS                       ESTHER                           

  middle_initial class_code  \
0              V   JD25       
1              V   3524       
2              J   JD25       
3              J   3524       
4                  4504       

                                         class_title        ethnicity  \
0  JUDGE, RETIRED                                ...  WHITE             
1  GENERAL COUNSEL IV                            ...  WHITE             
2  JUDGE, RETIRED                                ...  WHITE             
3  GENERAL COUNSEL IV                            ...  WHITE             
4  CORREC  OFFICER IV                            ...  HISPANIC          

            gender                                    status employ_date  \
0  MALE             URP - UNCLASSIFIED REGULAR PART-TIME        02/18/88   
1  MALE             CTP - CLASSIFIED TEMPORARY PART-TIME        02/01/15   
2  MALE             URP - UNCLASSIFIED REGULAR PART-TIME        02/01/20   
3  MALE             CTP - CLASSIFIED TEMPORARY PART-TIME        09/01/18   
4  FEMALE           CRF - CLASSIFIED REGULAR FULL-TIME          06/29/20   

   hourly_rate  hours_per_week  monthly_income  annual_income  state_number  \
0     75.96150            29.0         9545.82      114549.84        127717   
1     81.04454             4.0         1404.77       16857.24        127717   
2     75.96150            29.0         9545.82      114549.84         59115   
3     81.04453             4.0         1404.77       16857.24         59115   
4      0.00000            40.0         3284.27       39411.24        165030   

  is_duplicated  multiple_full_time_jobs combined_multiple_jobs  \
0          True                      NaN                    NaN   
1          True                      NaN                    NaN   
2          True                      NaN                    NaN   
3          True                      NaN                    NaN   
4          True                      1.0                    NaN   

   summed_annual_salary hide_from_search  
0             131407.08              NaN  
1                   NaN             True  
2             131407.08              NaN  
3                   NaN             True  
4                   NaN              NaN

In [7]:
data.tail()

agency                                        agency_name  \
149476     809  STATE PRESERVATION BOARD                      ...   
149477     809  STATE PRESERVATION BOARD                      ...   
149478     809  STATE PRESERVATION BOARD                      ...   
149479     809  STATE PRESERVATION BOARD                      ...   
149480     809  STATE PRESERVATION BOARD                      ...   

                             last_name                      first_name  \
149476  WESSELS                         JOHN                             
149477  WINDHAM                         EVAN                             
149478  WRIGHT                          DERRICK                          
149479  YOUNG                           DOUGLAS                          
149480  ZUNKER                          GEORGIA                          

       middle_initial class_code  \
149476              P   6232       
149477              A   0302       
149478              C   0130       
149479              R   1572       
149480              P   0130       

                                              class_title        ethnicity  \
149476  SECURITY OFFICER III                          ...  WHITE             
149477  WEB ADMINISTRATOR III                         ...  WHITE             
149478  CUSTOMER SERVICE REP I                        ...  WHITE             
149479  PROGRAM SPECIALIST III                        ...  WHITE             
149480  CUSTOMER SERVICE REP I                        ...  WHITE             

                 gender                                    status employ_date  \
149476  MALE             CRF - CLASSIFIED REGULAR FULL-TIME          10/30/17   
149477  FEMALE           CRF - CLASSIFIED REGULAR FULL-TIME          07/13/15   
149478  MALE             CRP - CLASSIFIED REGULAR PART-TIME          10/15/12   
149479  MALE             CRF - CLASSIFIED REGULAR FULL-TIME          09/22/89   
149480  FEMALE           CRP - CLASSIFIED REGULAR PART-TIME          02/16/12   

        hourly_rate  hours_per_week  monthly_income  annual_income  \
149476         0.00            40.0         2899.00       34788.00   
149477         0.00            40.0         5500.00       66000.00   
149478        12.93            20.0         1120.60       13447.20   
149479         0.00            40.0         5744.16       68929.92   
149480        11.74            20.0         1017.46       12209.52   

        state_number is_duplicated  multiple_full_time_jobs  \
149476        770781           NaN                      NaN   
149477        847431           NaN                      NaN   
149478         34266           NaN                      NaN   
149479        123490           NaN                      NaN   
149480        103583           NaN                      NaN   

       combined_multiple_jobs  summed_annual_salary hide_from_search  
149476                    NaN                   NaN              NaN  
149477                    NaN                   NaN              NaN  
149478                    NaN                   NaN              NaN  
149479                    NaN                   NaN              NaN  
149480                    NaN                   NaN              NaN

In [8]:
data.shape

(149481, 21)

#### There are 149481 rows and 21 columns in the data.

In [9]:
# Checking for True duplicates
data.duplicated().sum

<bound method Series.sum of 0         False
1         False
2         False
3         False
4         False
          ...  
149476    False
149477    False
149478    False
149479    False
149480    False
Length: 149481, dtype: bool>

#### There are no actual Duplicates(True Duplicates) in the data

#### In the dataset, the column is_duplicated has values like "True" or "NoN". values marked as "True" indicate that the row is considered duplicated based on at least one column. This means that these rows are not true duplicates with identical entries in every column.

In [12]:
# Displaying the unique values in the "is_duplicated" column.
print(data['is_duplicated'].unique())
# Displaying count of values other then NaN
print(data['is_duplicated'].value_counts())

[True nan]
is_duplicated
True    143
Name: count, dtype: int64


#### So there are only two types of values present in the column 'is_duplicated'

#### So there are 143 rows where the data is duplicates based on at least one column.

In [15]:
# Displaying unique values in the "multiple_full_time_jobs" column.
print(data['multiple_full_time_jobs'].unique())
# Displaying count of values other than NaN
print(data['multiple_full_time_jobs'].value_counts())

[nan  1.]
multiple_full_time_jobs
1.0    14
Name: count, dtype: int64


#### There are two types of values in the 'multiple_full_time_jobs' column i.e 'NaN' and '1' and the '1' value is also very less that is only 14 times.

In [17]:
# Displaying unique values in the "combined_multiple_jobs" column.
print(data['combined_multiple_jobs'].unique())
# Displaying count of values other than NaN
print(data['combined_multiple_jobs'].value_counts())

[nan True]
combined_multiple_jobs
True    97
Name: count, dtype: int64


#### There are two type of values in the "combined_multiple_jobs" column 'True' and 'NaN' with count of 'True' values is only 97.

In [30]:
# Displaying unique values in the "summed_anual_salary" column.
print(data['summed_annual_salary'].unique())
# Displaying count of values other than NaN
print(data['summed_annual_salary'].value_counts())

[131407.08       nan 171383.4   31950.    33300.    33966.    39174.
  35570.52  56952.6   21839.76 121200.    90000.   150000.    92400.
  18000.    30312.84]
summed_annual_salary
131407.08    2
171383.40    1
31950.00     1
33300.00     1
33966.00     1
39174.00     1
35570.52     1
56952.60     1
21839.76     1
121200.00    1
90000.00     1
150000.00    1
92400.00     1
18000.00     1
30312.84     1
Name: count, dtype: int64


#### There are 16 unique values in the 'summed_annual_salary' column and each value is repeated only once and the value '131407.08' is repeated two times and all other values are 'NaN'.

In [33]:
# Displaying unique values in the "hide_from_search" column.
print(data['hide_from_search'].unique())
# Displaying count of values other than NaN
print(data['hide_from_search'].value_counts())

[nan True]
hide_from_search
True    16
Name: count, dtype: int64


#### There are only two types of values present in the 'hide_from_search_column' which are 'True' and 'NaN' and the 'True' value is repeated only 16 times.

### Overall onclusion based on the above four columns :
#### The columns 'is_duplicated', 'multiple_full_time_jobs', 'combined_multiple_jobs', 'summed_annual_salary', 'hide_from_search' not contribute not much in the dataset because there are most of the values are 'NaN'.
#### So we remove these columns from the dataset.

In [37]:
# List of columns to remove
columns_to_remove = ['is_duplicated',  'multiple_full_time_jobs', 'combined_multiple_jobs', 
                     'summed_annual_salary', 'hide_from_search']

# Removing the specified columns
data_cleaned = data.drop(columns=columns_to_remove)
data_cleaned

agency                                        agency_name  \
0          241  COMPTROLLER OF PUBLIC ACCOUNTS, JUDICIARY SECT...   
1          212  OFFICE OF COURT ADMINISTRATION                ...   
2          241  COMPTROLLER OF PUBLIC ACCOUNTS, JUDICIARY SECT...   
3          212  OFFICE OF COURT ADMINISTRATION                ...   
4          696  TEXAS DEPARTMENT OF CRIMINAL JUSTICE          ...   
...        ...                                                ...   
149476     809  STATE PRESERVATION BOARD                      ...   
149477     809  STATE PRESERVATION BOARD                      ...   
149478     809  STATE PRESERVATION BOARD                      ...   
149479     809  STATE PRESERVATION BOARD                      ...   
149480     809  STATE PRESERVATION BOARD                      ...   

                             last_name                      first_name  \
0       RUCKER                          MORTON                           
1       RUCKER                          MORTON                           
2       SPECIA JR                       JOHN                             
3       SPECIA JR                       JOHN                             
4       ONTIVEROS                       ESTHER                           
...                                ...                             ...   
149476  WESSELS                         JOHN                             
149477  WINDHAM                         EVAN                             
149478  WRIGHT                          DERRICK                          
149479  YOUNG                           DOUGLAS                          
149480  ZUNKER                          GEORGIA                          

       middle_initial class_code  \
0                   V   JD25       
1                   V   3524       
2                   J   JD25       
3                   J   3524       
4                       4504       
...               ...        ...   
149476              P   6232       
149477              A   0302       
149478              C   0130       
149479              R   1572       
149480              P   0130       

                                              class_title        ethnicity  \
0       JUDGE, RETIRED                                ...  WHITE             
1       GENERAL COUNSEL IV                            ...  WHITE             
2       JUDGE, RETIRED                                ...  WHITE             
3       GENERAL COUNSEL IV                            ...  WHITE             
4       CORREC  OFFICER IV                            ...  HISPANIC          
...                                                   ...              ...   
149476  SECURITY OFFICER III                          ...  WHITE             
149477  WEB ADMINISTRATOR III                         ...  WHITE             
149478  CUSTOMER SERVICE REP I                        ...  WHITE             
149479  PROGRAM SPECIALIST III                        ...  WHITE             
149480  CUSTOMER SERVICE REP I                        ...  WHITE             

                 gender                                    status employ_date  \
0       MALE             URP - UNCLASSIFIED REGULAR PART-TIME        02/18/88   
1       MALE             CTP - CLASSIFIED TEMPORARY PART-TIME        02/01/15   
2       MALE             URP - UNCLASSIFIED REGULAR PART-TIME        02/01/20   
3       MALE             CTP - CLASSIFIED TEMPORARY PART-TIME        09/01/18   
4       FEMALE           CRF - CLASSIFIED REGULAR FULL-TIME          06/29/20   
...                 ...                                       ...         ...   
149476  MALE             CRF - CLASSIFIED REGULAR FULL-TIME          10/30/17   
149477  FEMALE           CRF - CLASSIFIED REGULAR FULL-TIME          07/13/15   
149478  MALE             CRP - CLASSIFIED REGULAR PART-TIME          10/15/12   
149479  MALE             CRF - CLASSIFIED REGULAR FULL-TIME          09/22/89   
149480  FEMALE           CRP 